internal/sshtunneler/tunneler.go

```diff
@@ -0,0 +1,259 @@
+	Dialer         SSHDial
+	Clock          clock.Clock
+}
+
+func (args *TrackerArgs) validate() error {
+	if args.State == nil {
+		return errors.New("state is required")
+	}
+	if args.ControllerInfo == nil {
+		return errors.New("controller info is required")
+	}
+	if args.Dialer == nil {
+		return errors.New("dialer is required")
+	}
+	if args.Clock == nil {
+		return errors.New("clock is required")
+	}
+	return nil
+}
+
+// NewTracker creates a new tunnel tracker.
+func NewTracker(args TrackerArgs) (*Tracker, error) {
+	if err := args.validate(); err != nil {
+		return nil, err
+	}
+
+	authn, err := newTunnelAuthentication(args.Clock)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Tracker{
+		tracker:    make(map[string]chan (net.Conn)),
+		authn:      authn,
+		controller: args.ControllerInfo,
+		clock:      args.Clock,
+		state:      args.State,
+		dialer:     args.Dialer,
+	}, nil
+}
+
+// RequestArgs holds the arguments for requesting a tunnel.
+type RequestArgs struct {
+	MachineID string
+	ModelUUID string
+}
+
+func (tt *Tracker) generateEphemeralSSHKey() (gossh.Signer, gossh.PublicKey, error) {
+	privKey, err := ssh.ED25519()
+	if err != nil {
+		return nil, nil, errors.Annotate(err, "failed to generate key")
+	}
+
+	sshPrivateKey, err := gossh.NewSignerFromKey(privKey)
+	if err != nil {
+		return nil, nil, errors.NotValidf("private key")
+	}
+
+	return sshPrivateKey, sshPrivateKey.PublicKey(), nil
+}
+
+// RequestTunnel requests a tunnel to a model specific unit.
+//
+// The returned tunnelRequest should be used to wait for the tunnel to be established.
+// See Wait() for more information.
+func (tt *Tracker) RequestTunnel(ctx context.Context, req RequestArgs) (*gossh.Client, error) {
+	tunnelID, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+
+	// We use the same expiry for the password and the state entry.
+	// The state's expiry is used to clean up any dangling requests.
+	// The password expiry is used to invalidate old passwords.
+	now := tt.clock.Now()
+
+	password, err := tt.authn.generatePassword(tunnelID.String(), now)
+	if err != nil {
+		return nil, err
+	}
+
+	privateKey, publicKey, err := tt.generateEphemeralSSHKey()
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure to use an unbuffered channel to ensure someone always
+	// has responsibility of the connection passed around.
+	connRecv := make(chan (net.Conn))
+
+	tt.add(tunnelID.String(), connRecv)
+	defer tt.delete(tunnelID.String())
+
+	args := state.SSHConnRequestArg{
+		TunnelID:            tunnelID.String(),
+		ModelUUID:           req.ModelUUID,
+		MachineId:           req.MachineID,
+		Expires:             now.Add(maxTimeout),
+		Username:            reverseTunnelUser,
+		Password:            password,
+		ControllerAddresses: tt.controller.Addresses(),
+		UnitPort:            0, // Allow the unit worker to determine the port.
+		EphemeralPublicKey:  publicKey.Marshal(),
+	}
+
+	err = tt.state.InsertSSHConnRequest(args)
```

Comment:
In a perfect world we would delete these requests when we return. But I think with the context changes below and the relatively short timeout window, I think it is fine.